locked-nft/contracts/NFTLocker.cdc

```diff
@@ -0,0 +1,176 @@
+                self.id = id
+                self.owner = lockedToken.owner
+                self.lockedAt = lockedToken.lockedAt
+                self.lockedUntil = lockedToken.lockedUntil
+                self.duration = lockedToken.duration
+                self.nftType = lockedToken.nftType
+            } else {
+                self.id = id
+                self.owner = owner
+                self.lockedAt = UInt64(getCurrentBlock().timestamp)
+                self.lockedUntil = self.lockedAt + duration
+                self.duration = duration
+                self.nftType = nftType
+            }
+        }
+    }
+
+    pub fun getNFTLockerDetails(id: UInt64, nftType: Type): NFTLocker.LockedData? {
+        return (NFTLocker.lockedTokens[nftType]!)[id]!
+    }
+
+    pub fun getLockedTokenKey(id: UInt64, nftType: Type): String {
+        return nftType.identifier.concat(".").concat(id.toString())
+    }
+
+    pub fun canUnlockToken(id: UInt64, nftType: Type): Bool {
+        if let lockedToken = (NFTLocker.lockedTokens[nftType]!)[id] {
+            if lockedToken.lockedUntil < UInt64(getCurrentBlock().timestamp) {
+                return true
+            }
+        }
+
+        return false
+    }
+
+    pub resource interface LockedCollection {
+        pub fun getIDs(): [String]
+        pub fun borrowNFT(key: String): &NonFungibleToken.NFT
+    }
+
+    pub resource interface LockProvider {
+        pub fun lock(token: @NonFungibleToken.NFT, duration: UInt64)
+        pub fun unlock(id: UInt64, nftType: Type): @NonFungibleToken.NFT
+    }
+
+    pub resource Collection: LockedCollection, LockProvider {
+        pub var lockedNFTs: @{String: NonFungibleToken.NFT}
+
+        pub fun unlock(id: UInt64, nftType: Type): @NonFungibleToken.NFT {
+            pre {
+                NFTLocker.canUnlockToken(
+                    id: id,
+                    nftType: nftType
+                ) == true : "locked duration has not been met"
+            }
+
+            let key: String = NFTLocker.getLockedTokenKey(id: id, nftType: nftType)
+            let token <- self.lockedNFTs.remove(key: key) ?? panic("Missing NFT")
+
+            if let lockedToken = NFTLocker.lockedTokens[nftType] {
+                lockedToken.remove(key: id)
+            }
+            NFTLocker.totalLockedTokens = NFTLocker.totalLockedTokens - 1
+
+            emit NFTUnlocked(
+                id: token.id,
+                from: self.owner?.address,
+                nftType: nftType
+            )
+
+            emit Withdraw(id: token.id, from: self.owner?.address)
+
+            return <-token
+        }
+
+        pub fun lock(token: @NonFungibleToken.NFT, duration: UInt64) {
+            let id: UInt64 = token.id
+            let nftType: Type = token.getType()
+            let key: String = NFTLocker.getLockedTokenKey(id: id, nftType: nftType)
+            let oldToken <- self.lockedNFTs[key] <- token
+
+            if NFTLocker.lockedTokens[nftType] == nil {
```

Comment:
Should this be removed since `nestedLock` is already defined as `NFTLocker.lockedTokens[nftType] ?? {}` (and the new updated `nestedLock` assigned to `NFTLocker.lockedTokens[nftType]` afterwards)?

Reply:
I left this, because I need it when forcing the get on lockedData. I did remove the ?? {} because it was duplicative. Nice catch.